### PR TITLE
nixos/ntpd: cleanup; add tests

### PR DESCRIPTION
**Pull request by pyrox0**
Brings the NTPd service more in line with modern services, reducing its security impact on a system, and also adds tests to ensure it works properly. Currently the tests are very basic, but they ensure that the unit starts properly and that userland tools are able to communicate with ntpd.

Part of #349572

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/networking/ntp/ntpd.nix [hunk 1]
```diff
@@ -132,6 +132,7 @@ in
 
         preStart =
           ''
+            # shellcheck disable=SC2174
             mkdir -m 0755 -p ${stateDir}
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/ntp/ntpd.nix:136]:
```suggestion
            mkdir -m 0755 ${stateDir} || true
```
what do you think about this?

Comment (Scrumplex) [reply nixos/modules/services/networking/ntp/ntpd.nix:136]: Perhaps this code can be removed completely by utilizing `StateDirectory` for the service. `stateDir` is hardcoded to be `/var/lib/ntp`

Comment (SuperSandro2000) [reply nixos/modules/services/networking/ntp/ntpd.nix:136]: Yeah, completely missed that!

Comment (pyrox0) [reply nixos/modules/services/networking/ntp/ntpd.nix:136]: Good call.

**Comment by pyrox0** [comment]
Fixed the suggested change, I'm going to submit a follow-up PR which increases the security of the service unit with systemd hardening.

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
Seems like there is no test for this module. I'll test this locally later

File: nixos/modules/services/networking/ntp/ntpd.nix [hunk 2]
```diff
@@ -113,34 +133,57 @@ in
...
 
-        serviceConfig = {
-          ExecStart = "@${ntp}/bin/ntpd ntpd -g ${builtins.toString ntpFlags}";
-          Type = "forking";
-        };
+    users.users.ntp = {
+      isSystemUser = true;
+      group = "ntp";
+      description = "NTP daemon user";
+      home = "/var/lib/ntp";
```
Comment (networkException) [inline nixos/modules/services/networking/ntp/ntpd.nix:144]: Can you remove the reference to the stateDir variable in the same commit as you removed it from the let in

This way it doesn't break eval for git bisect

Comment (pyrox0) [reply nixos/modules/services/networking/ntp/ntpd.nix:144]: Done!

**Comment by Scrumplex** [comment]
@ofborg test ntpd

**Review by Scrumplex — COMMENTED** [review]
Changes LGTM.

Should we additionally run ntpd as a non-root user?

**Comment by pyrox0** [comment]
> Changes LGTM.
> 
> Should we additionally run ntpd as a non-root user?

Unfortunately, that's not possible. For some reason, NTPd does not support running as a non-root user, but with the `-u ntp:ntp` flag(lines 34-35) will drop privilages as soon as possible. I've searched for patches that would disable the root check if appropriate capabilities are given to the service, but nothing I've done worked. Therefore, it can't be, at least for now. Since it does drop privilages quickly, there's less risk, but it would still be nice if this could be patched in the future.

**Review by SuperSandro2000 — APPROVED** [review]
Lgtm

**Review by Scrumplex — APPROVED** [review]

**Pull request merged**